Add content from community doc PR 421 (#139)
rancher/k3s-product-docs: versions/latest/modules/en/pages/upgrades/roll-back.adoc

= Rolling Back K3s

You can roll back the K3s Kubernetes version after an upgrade, using a combination of K3s binary downgrade and datastore restoration. Rollback can be performed on clusters of all types, including a single-node SQLite, an external datastore, or an embedded etcd. When rolling back to a previous Kubernetes minor version, you must have a datastore snapshot taken on the Kubernetes minor version you wish to roll back to.

[WARNING]
====

> If you cannot restore the database, you can not roll back to a previous minor version.

Reviewer: ```suggestion
If you cannot restore the database, you cannot roll back to a previous minor version.
```

====

== Important Considerations

* *Backups:* Before upgrading, ensure you have a valid database or etcd snapshot from your cluster running the older version of K3s. Without a backup, a rollback is impossible.
* *Potential Data Loss:* The `k3s-killall.sh` script forcefully terminates K3s processes and may result in data loss if applications are not properly shut down.
* *Version Specifics:* Always verify K3s and component versions before and after the rollback.

== Rolling Back a K3s Cluster

[tabs]

> =====

Author: @btat Thanks for catching this.

SQLite::

> +

Reviewer: Can you ensure all tab item contents are enclosed within `--`.

E..g

```
Name of tab::
+
--
<insert-tab-contents>
--
```

I don't think the syntax is correct even though it renders fine. Perhaps ending a tab with a list is special, but as soon as the tab content ends with something else (paragraph, admonition, etc.) it breaks.

E.g. 

This is fine:

```
Embedded etcd::
+
To roll back a K3s cluster when using an embedded etcd, follow these steps: 

<list>
```

But this is not:

```
Embedded etcd::
+
To roll back a K3s cluster when using an embedded etcd, follow these steps: 

<list>

This is the conclusion paragraph for the tab.
```

Author: @btat thanks for clarifying this. Updated via 17aed83eed4703a030b74417ef08dc7e47904b5e

k3s-killall.sh
----

. On each node, roll back the K3s binary to the previous version, but *do not* start K3s.

* Clusters with Internet Access:

** Server nodes: 
+
[,bash]
----
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=vX.Y.Zk3s1 INSTALL_K3S_EXEC="server" INSTALL_K3S_SKIP_START="true" sh -
----

** Agent nodes:
+
[,bash]
----
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=vX.Y.Zk3s1 INSTALL_K3S_EXEC="agent" INSTALL_K3S_SKIP_START="true" sh -
----

* Air-gapped Clusters:

** Download the artifacts and run the xref:installation/airgap.adoc#_install_k3s[install script] locally. Add the environment variable `INSTALL_K3S_SKIP_START="true"` when running the install script to prevent K3s from starting.

. On the first server node or the node without a `server:` entry in its xref:installation/configuration.adoc[K3s config file], initiate the cluster restore. Refer to the xref:cli/etcd-snapshot.adoc#_snapshot_restore_steps[Snapshot Restore Steps] for more information: 
+
[,bash]
----
k3s server --cluster-reset --cluster-reset-restore-path=<PATH-TO-SNAPSHOT>
----
+
[WARNING]
====

> This will overwrite all data in the etcd datastore. Verify the snapshot's integrity before restoring. Be aware that large snapshots can take a long time to restore.

Reviewer: ```suggestion
This overwrites all data in the etcd datastore. Verify the snapshot's integrity before restoring. Be aware that large snapshots can take a long time to restore.
```

External Database:: 
+
To roll back a K3s cluster when using an external database (e.g., PostgreSQL, MySQL), follow these steps: 

. If the cluster is running and the Kubernetes API is available, gracefully stop workloads by draining all nodes: 
+
[,bash]
----
kubectl drain --ignore-daemonsets --delete-emptydir-data <NODE-ONE-NAME> <NODE-TWO-NAME> <NODE-THREE-NAME> ...
----
+
[NOTE]
====
This process may disrupt running applications.
====

. On each node, stop the K3s service and all running pod processes:
+
[,bash]
----
k3s-killall.sh
----

. Restore a database snapshot taken before upgrading K3s and verify the integrity of the database. For example, if you're using PostgreSQL, run the following command:
+
[,bash]
----
pg_restore -U <DB-USER> -d <DB-NAME> <BACKUP-FILE>
----

. On each node, roll back the K3s binary to the previous version. 

* Clusters with Internet Access:

** Server nodes:
+
[,bash]
----
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=vX.Y.Zk3s1 INSTALL_K3S_EXEC="server" sh -
----

** Agent nodes:
+
[,bash]
----
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=vX.Y.Zk3s1 INSTALL_K3S_EXEC="agent" sh -
----

* Air-gapped Clusters:

> ** Download the artifacts and run the xref:installation/airgap.adoc#_install_k3s[install script] locally. Verify the K3s version after install with `k3s --version` and reapply any custom configurations that where used before the upgrade. 

Reviewer: ```suggestion
** Download the artifacts and run the xref:installation/airgap.adoc#_install_k3s[install script] locally. Verify the K3s version after install with `k3s --version` and reapply any custom configurations that were used before the upgrade. 
```
